<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span></li><li><span><a href="#Создание-моделей" data-toc-modified-id="Создание-моделей-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Создание моделей</a></span><ul class="toc-item"><li><span><a href="#Модель-логистической-регрессии" data-toc-modified-id="Модель-логистической-регрессии-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Модель логистической регрессии</a></span></li><li><span><a href="#Модель-решающего-дерева" data-toc-modified-id="Модель-решающего-дерева-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Модель решающего дерева</a></span></li><li><span><a href="#Модель-случайного-леса" data-toc-modified-id="Модель-случайного-леса-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Модель случайного леса</a></span></li></ul></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><span><a href="#Модель-логистической-регрессии" data-toc-modified-id="Модель-логистической-регрессии-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Модель логистической регрессии</a></span></li><li><span><a href="#Модель-решающего-дерева" data-toc-modified-id="Модель-решающего-дерева-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Модель решающего дерева</a></span></li><li><span><a href="#Модель-случайного-леса" data-toc-modified-id="Модель-случайного-леса-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Модель случайного леса</a></span></li><li><span><a href="#Изменение-порогов-логистической-регрессии" data-toc-modified-id="Изменение-порогов-логистической-регрессии-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Изменение порогов логистической регрессии</a></span></li><li><span><a href="#Увеличение-выборки" data-toc-modified-id="Увеличение-выборки-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Увеличение выборки</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.5.1"><span class="toc-item-num">3.5.1&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Решающее-дерево" data-toc-modified-id="Решающее-дерево-3.5.2"><span class="toc-item-num">3.5.2&nbsp;&nbsp;</span>Решающее дерево</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-3.5.3"><span class="toc-item-num">3.5.3&nbsp;&nbsp;</span>Случайный лес</a></span></li></ul></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Модель-решающего-дерева" data-toc-modified-id="Модель-решающего-дерева-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Модель решающего дерева</a></span></li><li><span><a href="#Модель-случайного-леса" data-toc-modified-id="Модель-случайного-леса-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Модель случайного леса</a></span></li><li><span><a href="#Проверим-модели-на-адекватность" data-toc-modified-id="Проверим-модели-на-адекватность-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Проверим модели на адекватность</a></span></li></ul></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Отток клиентов

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. Целевая метрика - F1 не менее 0.59
Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

## Подготовка данных

In [1]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import OrdinalEncoder
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import f1_score
from sklearn.metrics import roc_auc_score
from sklearn.dummy import DummyClassifier
from sklearn.utils import shuffle
from sklearn.metrics import confusion_matrix
pd.options.mode.chained_assignment = None

data = pd.read_csv('/datasets/Churn.csv')
display(data.shape)
display(data.info())
data.head(10)

(10000, 14)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
RowNumber          10000 non-null int64
CustomerId         10000 non-null int64
Surname            10000 non-null object
CreditScore        10000 non-null int64
Geography          10000 non-null object
Gender             10000 non-null object
Age                10000 non-null int64
Tenure             9091 non-null float64
Balance            10000 non-null float64
NumOfProducts      10000 non-null int64
HasCrCard          10000 non-null int64
IsActiveMember     10000 non-null int64
EstimatedSalary    10000 non-null float64
Exited             10000 non-null int64
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


None

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   
5          6    15574012       Chu          645     Spain    Male   44   
6          7    15592531  Bartlett          822    France    Male   50   
7          8    15656148    Obinna          376   Germany  Female   29   
8          9    15792365        He          501    France    Male   44   
9         10    15592389        H?          684    France    Male   27   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   
5     8.0  113755.78              2          1               0   
6     7.0       0.00              2          1               1   
7     4.0  115046.74              4          1               0   
8     4.0  142051.07              2          0               1   
9     2.0  134603.88              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0  
5        149756.71       1  
6         10062.80       0  
7        119346.88       1  
8         74940.50       0  
9         71725.73       0

Есть пропуски в Tenure. Так как заполнить пропуски по имеющимся данным без искажения нельзя, удалим их. Также приведем названия столбцов к нижнему регистру.

In [2]:
data.columns = data.columns.str.lower() #приведем названия к нижнему регистру
data = data.dropna()
display(data.info())
data.head(10)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9091 entries, 0 to 9998
Data columns (total 14 columns):
rownumber          9091 non-null int64
customerid         9091 non-null int64
surname            9091 non-null object
creditscore        9091 non-null int64
geography          9091 non-null object
gender             9091 non-null object
age                9091 non-null int64
tenure             9091 non-null float64
balance            9091 non-null float64
numofproducts      9091 non-null int64
hascrcard          9091 non-null int64
isactivemember     9091 non-null int64
estimatedsalary    9091 non-null float64
exited             9091 non-null int64
dtypes: float64(3), int64(8), object(3)
memory usage: 1.0+ MB


None

rownumber  customerid   surname  creditscore geography  gender  age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   
5          6    15574012       Chu          645     Spain    Male   44   
6          7    15592531  Bartlett          822    France    Male   50   
7          8    15656148    Obinna          376   Germany  Female   29   
8          9    15792365        He          501    France    Male   44   
9         10    15592389        H?          684    France    Male   27   

   tenure    balance  numofproducts  hascrcard  isactivemember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   
5     8.0  113755.78              2          1               0   
6     7.0       0.00              2          1               1   
7     4.0  115046.74              4          1               0   
8     4.0  142051.07              2          0               1   
9     2.0  134603.88              1          1               1   

   estimatedsalary  exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0  
5        149756.71       1  
6         10062.80       0  
7        119346.88       1  
8         74940.50       0  
9         71725.73       0

In [3]:
for column in data.columns[3:]:
    display (column, data[column].unique())

'creditscore'

array([619, 608, 502, 699, 850, 645, 822, 376, 501, 684, 528, 497, 476,
       549, 635, 616, 653, 587, 726, 732, 636, 510, 669, 846, 577, 756,
       571, 574, 411, 533, 553, 520, 722, 475, 490, 804, 582, 472, 465,
       556, 834, 660, 776, 829, 637, 698, 788, 601, 656, 725, 511, 614,
       687, 555, 603, 751, 581, 735, 661, 675, 738, 813, 657, 604, 519,
       664, 678, 757, 416, 665, 777, 506, 493, 750, 729, 646, 647, 808,
       524, 769, 515, 773, 814, 710, 623, 670, 622, 785, 605, 479, 685,
       562, 721, 628, 668, 828, 674, 625, 770, 758, 795, 686, 789, 589,
       461, 584, 579, 663, 682, 793, 691, 485, 650, 754, 535, 716, 539,
       706, 586, 631, 717, 683, 704, 615, 667, 484, 480, 512, 606, 597,
       778, 514, 730, 525, 715, 580, 807, 521, 759, 516, 711, 550, 618,
       643, 671, 689, 620, 676, 572, 695, 592, 694, 547, 594, 673, 610,
       767, 585, 763, 712, 703, 662, 659, 523, 772, 545, 634, 739, 771,
       681, 544, 766, 727, 557, 531, 498, 651, 791, 733, 811, 70

'geography'

array(['France', 'Spain', 'Germany'], dtype=object)

'gender'

array(['Female', 'Male'], dtype=object)

'age'

array([42, 41, 39, 43, 44, 50, 29, 27, 31, 24, 34, 25, 35, 45, 58, 32, 38,
       46, 36, 33, 40, 51, 61, 49, 37, 19, 66, 56, 26, 21, 55, 22, 30, 28,
       65, 52, 48, 57, 73, 47, 54, 72, 20, 67, 79, 62, 53, 80, 59, 68, 75,
       23, 60, 63, 64, 70, 18, 82, 69, 74, 71, 76, 77, 88, 85, 84, 81, 78,
       92, 83])

'tenure'

array([ 2.,  1.,  8.,  7.,  4.,  6.,  3., 10.,  5.,  9.,  0.])

'balance'

array([     0.  ,  83807.86, 159660.8 , ..., 155060.41,  57369.61,
        75075.31])

'numofproducts'

array([1, 3, 2, 4])

'hascrcard'

array([1, 0])

'isactivemember'

array([1, 0])

'estimatedsalary'

array([101348.88, 112542.58, 113931.57, ..., 101699.77,  42085.58,
        92888.52])

'exited'

array([1, 0])

Удалим rownumber, customerid, surname, так как очевидно, что эти поля не влияют на целевой признак.

In [4]:
data_ohe = data.drop(columns=['rownumber','customerid', 'surname'], axis =1)
data_ohe

creditscore geography  gender  age  tenure    balance  numofproducts  \
0             619    France  Female   42     2.0       0.00              1   
1             608     Spain  Female   41     1.0   83807.86              1   
2             502    France  Female   42     8.0  159660.80              3   
3             699    France  Female   39     1.0       0.00              2   
4             850     Spain  Female   43     2.0  125510.82              1   
...           ...       ...     ...  ...     ...        ...            ...   
9994          800    France  Female   29     2.0       0.00              2   
9995          771    France    Male   39     5.0       0.00              2   
9996          516    France    Male   35    10.0   57369.61              1   
9997          709    France  Female   36     7.0       0.00              1   
9998          772   Germany    Male   42     3.0   75075.31              2   

      hascrcard  isactivemember  estimatedsalary  exited  
0             1               1        101348.88       1  
1             0               1        112542.58       0  
2             1               0        113931.57       1  
3             0               0         93826.63       0  
4             1               1         79084.10       0  
...         ...             ...              ...     ...  
9994          0               0        167773.55       0  
9995          1               0         96270.64       0  
9996          1               1        101699.77       0  
9997          0               1         42085.58       1  
9998          1               0         92888.52       1  

[9091 rows x 11 columns]

Для использования моделей логистической регрессии, решающего дерева и случайного леса категориальные данные преобразуем в численные с помощью порядкового One-Hot Encoding.

In [5]:
data_ohe = pd.get_dummies(data_ohe,drop_first = True)
data_ohe.shape
data_ohe

creditscore  age  tenure    balance  numofproducts  hascrcard  \
0             619   42     2.0       0.00              1          1   
1             608   41     1.0   83807.86              1          0   
2             502   42     8.0  159660.80              3          1   
3             699   39     1.0       0.00              2          0   
4             850   43     2.0  125510.82              1          1   
...           ...  ...     ...        ...            ...        ...   
9994          800   29     2.0       0.00              2          0   
9995          771   39     5.0       0.00              2          1   
9996          516   35    10.0   57369.61              1          1   
9997          709   36     7.0       0.00              1          0   
9998          772   42     3.0   75075.31              2          1   

      isactivemember  estimatedsalary  exited  geography_Germany  \
0                  1        101348.88       1                  0   
1                  1        112542.58       0                  0   
2                  0        113931.57       1                  0   
3                  0         93826.63       0                  0   
4                  1         79084.10       0                  0   
...              ...              ...     ...                ...   
9994               0        167773.55       0                  0   
9995               0         96270.64       0                  0   
9996               1        101699.77       0                  0   
9997               1         42085.58       1                  0   
9998               0         92888.52       1                  1   

      geography_Spain  gender_Male  
0                   0            0  
1                   1            0  
2                   0            0  
3                   0            0  
4                   1            0  
...               ...          ...  
9994                0            0  
9995                0            1  
9996                0            1  
9997                0            0  
9998                0            1  

[9091 rows x 12 columns]

In [6]:
for column in data_ohe.columns:
    display (column, data_ohe[column].unique())

'creditscore'

array([619, 608, 502, 699, 850, 645, 822, 376, 501, 684, 528, 497, 476,
       549, 635, 616, 653, 587, 726, 732, 636, 510, 669, 846, 577, 756,
       571, 574, 411, 533, 553, 520, 722, 475, 490, 804, 582, 472, 465,
       556, 834, 660, 776, 829, 637, 698, 788, 601, 656, 725, 511, 614,
       687, 555, 603, 751, 581, 735, 661, 675, 738, 813, 657, 604, 519,
       664, 678, 757, 416, 665, 777, 506, 493, 750, 729, 646, 647, 808,
       524, 769, 515, 773, 814, 710, 623, 670, 622, 785, 605, 479, 685,
       562, 721, 628, 668, 828, 674, 625, 770, 758, 795, 686, 789, 589,
       461, 584, 579, 663, 682, 793, 691, 485, 650, 754, 535, 716, 539,
       706, 586, 631, 717, 683, 704, 615, 667, 484, 480, 512, 606, 597,
       778, 514, 730, 525, 715, 580, 807, 521, 759, 516, 711, 550, 618,
       643, 671, 689, 620, 676, 572, 695, 592, 694, 547, 594, 673, 610,
       767, 585, 763, 712, 703, 662, 659, 523, 772, 545, 634, 739, 771,
       681, 544, 766, 727, 557, 531, 498, 651, 791, 733, 811, 70

'age'

array([42, 41, 39, 43, 44, 50, 29, 27, 31, 24, 34, 25, 35, 45, 58, 32, 38,
       46, 36, 33, 40, 51, 61, 49, 37, 19, 66, 56, 26, 21, 55, 22, 30, 28,
       65, 52, 48, 57, 73, 47, 54, 72, 20, 67, 79, 62, 53, 80, 59, 68, 75,
       23, 60, 63, 64, 70, 18, 82, 69, 74, 71, 76, 77, 88, 85, 84, 81, 78,
       92, 83])

'tenure'

array([ 2.,  1.,  8.,  7.,  4.,  6.,  3., 10.,  5.,  9.,  0.])

'balance'

array([     0.  ,  83807.86, 159660.8 , ..., 155060.41,  57369.61,
        75075.31])

'numofproducts'

array([1, 3, 2, 4])

'hascrcard'

array([1, 0])

'isactivemember'

array([1, 0])

'estimatedsalary'

array([101348.88, 112542.58, 113931.57, ..., 101699.77,  42085.58,
        92888.52])

'exited'

array([1, 0])

'geography_Germany'

array([0, 1], dtype=uint64)

'geography_Spain'

array([0, 1], dtype=uint64)

'gender_Male'

array([0, 1], dtype=uint64)

Отделим целевой признак от других признаков, выделим обучающую, валидационноую и тестовые выборки.

In [7]:
features = data_ohe.drop(columns='exited', axis =1)
target = data_ohe['exited']

features_train, features_valid_test, target_train, target_valid_test = train_test_split(features, target, test_size=0.4, random_state=12345)
features_valid, features_test, target_valid, target_test = train_test_split(features_valid_test, target_valid_test, test_size=0.5, random_state=12345)
massives = [features_train,target_train, features_valid, target_valid, features_test, target_test]
for massive in massives:
    display (massive.shape)

(5454, 11)

(5454,)

(1818, 11)

(1818,)

(1819, 11)

(1819,)

Проведем стандартизацию данных.

In [8]:
numeric = ['creditscore', 'age', 'tenure', 'balance', 'numofproducts', 'estimatedsalary']
scaler = StandardScaler()
scaler.fit(features_train[numeric])
features_train[numeric] = scaler.transform(features_train[numeric])
features_valid[numeric] = scaler.transform(features_valid[numeric])
features_test[numeric] = scaler.transform(features_test[numeric])


features_train = pd.DataFrame(features_train, index=target_train.index, columns=data_ohe.drop(columns='exited', axis =1).columns)
features_valid = pd.DataFrame(features_valid, index=target_valid.index, columns=data_ohe.drop(columns='exited', axis =1).columns)
features_test = pd.DataFrame(features_test, index=target_test.index, columns=data_ohe.drop(columns='exited', axis =1).columns)

display(features_train)
display(features_valid)
display(features_test)

creditscore       age    tenure   balance  numofproducts  hascrcard  \
9344     0.809075 -1.039327 -1.025995  0.554904      -0.908179          1   
3796    -1.152518 -1.227561  0.696524  0.480609      -0.908179          0   
7462    -0.398853  0.090079  1.385532 -1.237830      -0.908179          1   
1508    -0.749875 -0.286389  0.352020 -1.237830       0.809300          1   
4478    -1.028628 -0.756975 -0.336987 -1.237830       0.809300          0   
...           ...       ...       ...       ...            ...        ...   
4938     1.283988 -0.474624  1.730036  0.813031      -0.908179          0   
4529     0.158652  0.466548  1.041028  1.071609      -0.908179          1   
3866    -1.049276 -0.945210  0.352020  0.449165      -0.908179          1   
2414    -0.956359  1.219485  1.385532 -1.237830      -0.908179          1   
5056     1.222043 -1.509912 -1.025995  1.015938      -0.908179          0   

      isactivemember  estimatedsalary  geography_Germany  geography_Spain  \
9344               0         0.019508                  0                0   
3796               0         0.056167                  0                0   
7462               1         0.848738                  0                0   
1508               1        -0.894953                  0                0   
4478               1        -1.284516                  0                0   
...              ...              ...                ...              ...   
4938               0        -1.092767                  1                0   
4529               1         0.260842                  0                0   
3866               1        -1.086965                  0                0   
2414               0         1.288527                  0                0   
5056               1        -1.244002                  1                0   

      gender_Male  
9344            0  
3796            1  
7462            1  
1508            1  
4478            1  
...           ...  
4938            1  
4529            0  
3866            0  
2414            0  
5056            1  

[5454 rows x 11 columns]

creditscore       age    tenure   balance  numofproducts  hascrcard  \
7445    -1.369326  0.560665 -0.336987 -1.237830      -0.908179          1   
8620     1.232367  0.090079  1.041028 -1.237830       0.809300          0   
1714     0.840048  0.560665  0.352020  1.231363      -0.908179          0   
5441     1.056856 -0.945210 -1.370498  0.951231      -0.908179          1   
9001     0.406433 -0.662858  0.352020  0.767800      -0.908179          1   
...           ...       ...       ...       ...            ...        ...   
7505     0.107031 -1.133444 -0.336987 -1.237830       0.809300          1   
2232     0.003790 -0.286389  1.041028 -1.237830       0.809300          1   
3720    -1.813265  0.090079  1.041028  1.226442       0.809300          1   
5196     0.726482 -1.604030  0.696524  0.610110      -0.908179          1   
2460     0.117356  0.184196 -1.370498  0.860215      -0.908179          1   

      isactivemember  estimatedsalary  geography_Germany  geography_Spain  \
7445               0        -0.086537                  0                0   
8620               1        -0.537457                  0                0   
1714               0         1.070393                  1                0   
5441               0        -0.576279                  0                0   
9001               1         0.662068                  0                1   
...              ...              ...                ...              ...   
7505               0        -0.016504                  0                1   
2232               0         1.056449                  0                1   
3720               0        -0.468713                  1                0   
5196               1         1.328951                  0                0   
2460               0        -1.278650                  1                0   

      gender_Male  
7445            0  
8620            0  
1714            1  
5441            1  
9001            1  
...           ...  
7505            1  
2232            1  
3720            1  
5196            0  
2460            1  

[1818 rows x 11 columns]

creditscore       age    tenure   balance  numofproducts  hascrcard  \
5170     1.707279 -0.756975 -0.336987 -1.237830       0.809300          1   
4180    -0.429826 -0.286389  1.730036  0.485538       0.809300          0   
7349    -0.171721  0.278313  0.352020 -0.269213       0.809300          0   
7469     0.385784 -0.380507  1.041028  0.464813      -0.908179          1   
3467    -1.142194  0.278313 -1.370498  0.353837       0.809300          0   
...           ...       ...       ...       ...            ...        ...   
8579     0.014114 -0.474624 -0.336987  0.474274      -0.908179          0   
7536     0.499350  0.278313 -1.370498 -1.237830      -0.908179          1   
2911     0.003790  1.219485  1.041028 -0.444590      -0.908179          1   
2789     0.581944  0.466548 -1.715002  1.742887      -0.908179          1   
7755     0.272218 -0.474624  1.041028 -1.237830       0.809300          1   

      isactivemember  estimatedsalary  geography_Germany  geography_Spain  \
5170               1         0.718362                  0                0   
4180               0         1.687305                  1                0   
7349               1         0.824128                  1                0   
7469               0        -1.118018                  0                0   
3467               1        -0.107642                  1                0   
...              ...              ...                ...              ...   
8579               1         0.734995                  0                0   
7536               0        -1.704943                  0                0   
2911               0        -1.006086                  0                0   
2789               0        -0.541247                  0                1   
7755               1         1.455393                  0                0   

      gender_Male  
5170            0  
4180            1  
7349            1  
7469            0  
3467            0  
...           ...  
8579            1  
7536            0  
2911            0  
2789            1  
7755            1  

[1819 rows x 11 columns]

## Создание моделей

Посмотрим соотношение ушедших и оставшихся клиентов

In [9]:
share = target.sum()/target.count()
share

0.2039379606203938

Доля ушедших - 20%. Обучим модели без изменения дисбаланса классов.

### Модель логистической регрессии

In [10]:
model = LogisticRegression(random_state=12345, solver='liblinear')
model.fit(features_train, target_train)
predicted_valid = model.predict(features_valid)
probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
print("F1:", f1_score(target_valid, predicted_valid), "AUC_ROC:", auc_roc)

F1: 0.30400000000000005 AUC_ROC: 0.7736191158144302


### Модель решающего дерева

In [11]:
best_model_tree = None
best_result = 0
best_depth = 0
best_auc_roc = 0
for depth in range(1, 11):
    model_DecisionTreeClassifier = DecisionTreeClassifier(random_state=12345, max_depth=depth)
    model_DecisionTreeClassifier.fit(features_train, target_train)
    predicted_valid = model_DecisionTreeClassifier.predict(features_valid)
    result= f1_score(target_valid, predicted_valid)
    probabilities_valid = model_DecisionTreeClassifier.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]
    auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    if result > best_result:
        best_model_tree = model_DecisionTreeClassifier
        best_result = result
        best_depth = depth
        best_auc_roc = auc_roc
print("Лучший F1:", best_result, "Максимальная глубина:", best_depth, "AUC_ROC:", best_auc_roc)

Лучший F1: 0.5764331210191083 Максимальная глубина: 7 AUC_ROC: 0.8346049843812412


/opt/conda/lib/python3.7/site-packages/sklearn/metrics/classification.py:1437: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)


### Модель случайного леса

In [12]:
best_model_RandomForestClassifier = None
best_result = 0
best_est = 0
best_depth = 0
best_auc_roc = 0
for est in range(10, 51, 10):
    for depth in range (1, 100):
        model_RandomForestClassifier = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth=depth)
        model_RandomForestClassifier.fit(features_train, target_train)
        predicted_valid = model_RandomForestClassifier.predict(features_valid) 
        result = f1_score(target_valid, predicted_valid)
        probabilities_valid = model_RandomForestClassifier.predict_proba(features_valid)
        probabilities_one_valid = probabilities_valid[:, 1]
        auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
        if result > best_result:
            best_model_RandomForestClassifier = model_RandomForestClassifier
            best_result = result
            best_est = est
            best_depth = depth
            best_auc_roc = auc_roc

print("Лучший F1:", best_result, "Количество деревьев:", best_est, "Максимальная глубина:", depth, "AUC_ROC:", best_auc_roc)

/opt/conda/lib/python3.7/site-packages/sklearn/metrics/classification.py:1437: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)
/opt/conda/lib/python3.7/site-packages/sklearn/metrics/classification.py:1437: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)
/opt/conda/lib/python3.7/site-packages/sklearn/metrics/classification.py:1437: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)
/opt/conda/lib/python3.7/site-packages/sklearn/metrics/classification.py:1437: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)
/opt/conda/lib/python3.7/site-packages/sklearn/metrics/classification.py:1437: UndefinedMetricWarning: F-sco

Лучший F1: 0.594855305466238 Количество деревьев: 40 Максимальная глубина: 99 AUC_ROC: 0.8539696681312897


## Борьба с дисбалансом

Сбалансируем классы

### Модель логистической регрессии

In [13]:
model_LogisticRegression = LogisticRegression(random_state=12345, solver='liblinear', class_weight='balanced')
model_LogisticRegression.fit(features_train, target_train)
predicted_valid = model_LogisticRegression.predict(features_valid)
probabilities_valid = model_LogisticRegression.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
print("F1:", f1_score(target_valid, predicted_valid), "AUC_ROC:", auc_roc)

F1: 0.509731232622799 AUC_ROC: 0.7777884132187896


### Модель решающего дерева

In [14]:
best_model_tree = None
best_f1_score = 0
best_depth = 0
for depth in range(1, 11):
    model_DecisionTreeClassifier = DecisionTreeClassifier(random_state=12345, max_depth=depth, class_weight='balanced')
    model_DecisionTreeClassifier.fit(features_train, target_train)
    predicted_valid = model_DecisionTreeClassifier.predict(features_valid)
    result= f1_score(target_valid, predicted_valid)
    probabilities_valid = model_DecisionTreeClassifier.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]
    auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    if result > best_f1_score:
        best_model_tree = model_DecisionTreeClassifier
        best_f1_score = result
        best_depth = depth
        best_auc_roc = auc_roc
print("Лучший F1:", best_result, "Максимальная глубина:", best_depth, "AUC_ROC:", best_auc_roc)

Лучший F1: 0.594855305466238 Максимальная глубина: 5 AUC_ROC: 0.8396523192522141


### Модель случайного леса

In [15]:
best_model_RandomForestClassifier = None
best_result = 0
best_est = 0
best_depth = 0
best_auc_roc = 0
for est in range(40, 51, 10):
    for depth in range (1, 100):
        model_RandomForestClassifier = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth=depth, class_weight='balanced')
        model_RandomForestClassifier.fit(features_train, target_train)
        predicted_valid = model_RandomForestClassifier.predict(features_valid) 
        result = f1_score(target_valid, predicted_valid)
        probabilities_valid = model_RandomForestClassifier.predict_proba(features_valid)
        probabilities_one_valid = probabilities_valid[:, 1]
        auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
        if result > best_result:
            best_model_RandomForestClassifier = model_RandomForestClassifier
            best_result = result
            best_est = est
            best_depth = depth
            best_auc_roc = auc_roc

print("Лучший F1:", best_result, "Количество деревьев:", best_est, "Максимальная глубина:", depth, "AUC_ROC:", best_auc_roc)

Лучший F1: 0.6442786069651741 Количество деревьев: 40 Максимальная глубина: 99 AUC_ROC: 0.8682170685329411


In [16]:
predicted_test = best_model_RandomForestClassifier.predict(features_test)
print("F1:", f1_score(target_test, predicted_test), "AUC_ROC:", roc_auc_score(target_test, predicted_test))
print(confusion_matrix(target_test,predicted_test))

F1: 0.6016042780748663 AUC_ROC: 0.7619297142457905
[[1296  172]
 [ 126  225]]


Качество модели логистической регрессии существенно повысилось, случайного леса улучшилось незначительно, решающего дерева ухудшилось. Найдем лучшую модель логистической регрессии изменяя пороги.

### Изменение порогов логистической регрессии

In [17]:
best_result = 0
best_threshold = 0
best_auc_roc = 0

for threshold in np.arange(0, 0.81, 0.02):
    predicted_valid = probabilities_one_valid > threshold
    result = f1_score(target_valid, predicted_valid)
    probabilities_valid = model_LogisticRegression.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]
    auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    if result > best_result:
            best_result = result
            best_threshold = threshold
            best_auc_roc = auc_roc
print("F1:", best_result, "Порог:", best_threshold, "AUC_ROC:", best_auc_roc)


F1: 0.5156576200417536 Порог: 0.54 AUC_ROC: 0.7777884132187896


Лучший F1 = 0.514 достигается при пороге 0.54, c AUC_ROC = 0,711 - это незначительно лучше, достигнутого уровня без изменения порога.

### Увеличение выборки
Увеличим долю объектов с положительным исходом, чтобы лучше обучить модель

In [18]:
def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    
    features_upsampled, target_upsampled = shuffle(
    features_upsampled, target_upsampled, random_state=12345)
    
    return features_upsampled, target_upsampled

features_upsampled, target_upsampled = upsample(features_train, target_train, 4)
display( "features_upsampled", features_upsampled.shape)
display("target_upsampled", target_upsampled.shape)

'features_upsampled'

(8832, 11)

'target_upsampled'

(8832,)

In [19]:
share = target_upsampled.sum()/target_upsampled.count()
share

0.509963768115942

#### Логистическая регрессия

In [20]:
model_LogisticRegression_upsampled = LogisticRegression(random_state=12345, solver='liblinear', class_weight='balanced')
model_LogisticRegression_upsampled.fit(features_upsampled, target_upsampled)
predicted_valid = model_LogisticRegression_upsampled.predict(features_valid)
probabilities_valid = model_LogisticRegression_upsampled.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
print("F1:", f1_score(target_valid, predicted_valid), "AUC_ROC:", auc_roc)

F1: 0.509731232622799 AUC_ROC: 0.7777552797294834


Результат не изменился. Значит имеющихся данных хватило модели для обучения.

#### Решающее дерево

In [21]:
best_model_tree_upsampled = None
best_f1_score = 0
best_depth = 0
for depth in range(1, 11):
    model_DecisionTreeClassifier = DecisionTreeClassifier(random_state=12345, max_depth=depth)
    model_DecisionTreeClassifier.fit(features_upsampled, target_upsampled)
    predicted_valid = model_DecisionTreeClassifier.predict(features_valid)
    result= f1_score(target_valid, predicted_valid)
    probabilities_valid = model_DecisionTreeClassifier.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]
    auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    if result > best_f1_score:
        best_model_tree_upsampled = model_DecisionTreeClassifier
        best_f1_score = result
        best_depth = depth
        best_auc_roc = auc_roc
print("Лучший F1:", best_result, "Максимальная глубина:", best_depth, "AUC_ROC:", best_auc_roc)

Лучший F1: 0.5156576200417536 Максимальная глубина: 5 AUC_ROC: 0.8396523192522141


Для данной модели upsample снизил F1 с 0,59 до 0,51, поэтому будем использовать версию, обученную на данных без upsample.

#### Случайный лес

In [22]:
best_model_RandomForestClassifier_upsampled = None
best_result = 0
best_est = 0
best_depth = 0
best_auc_roc = 0
for est in range(40, 51, 10):
    for depth in range (1, 100):
        model_RandomForestClassifier = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth=depth, class_weight='balanced')
        model_RandomForestClassifier.fit(features_upsampled, target_upsampled)
        predicted_valid = model_RandomForestClassifier.predict(features_valid) 
        result = f1_score(target_valid, predicted_valid)
        probabilities_valid = model_RandomForestClassifier.predict_proba(features_valid)
        probabilities_one_valid = probabilities_valid[:, 1]
        auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
        if result > best_result:
            best_model_RandomForestClassifier_upsampled = model_RandomForestClassifier
            best_result = result
            best_est = est
            best_depth = depth
            best_auc_roc = auc_roc

print("Лучший F1:", best_result, "Количество деревьев:", best_est, "Максимальная глубина:", depth, "AUC_ROC:", best_auc_roc)

Лучший F1: 0.6304347826086956 Количество деревьев: 40 Максимальная глубина: 99 AUC_ROC: 0.8630371997047439


Метрика F1 модели на увеличенных данных несколько хуже: 0,63 против 0,64.

## Тестирование модели

Проведем тестирование моделей на тестовых данных.

### Логистическая регрессия

In [33]:
probabilities_test = model_LogisticRegression.predict_proba(features_test)
probabilities_one_test = probabilities_test[:, 1]

predicted_test = probabilities_one_test > best_threshold
print("F1:", f1_score(target_test, predicted_test), "Порог:", best_threshold, "AUC_ROC:",
      roc_auc_score(target_test, probabilities_one_test))
print(confusion_matrix(target_test,predicted_test))

F1: 0.4916201117318436 Порог: 0.54 AUC_ROC: 0.7817621121435836
[[1144  324]
 [ 131  220]]


Модель практически не переобучена, F1 снизилась с 0,5 до 0,49 

### Модель решающего дерева

In [31]:
predicted_test = best_model_tree.predict(features_test)
probabilities_test = best_model_tree.predict_proba(features_test)
probabilities_one_test = probabilities_test[:, 1]
auc_roc = roc_auc_score(target_test, probabilities_one_test)
print("F1:", f1_score(target_test, predicted_test), "AUC_ROC:", auc_roc)
print(confusion_matrix(target_test,predicted_test))

F1: 0.5653631284916201 AUC_ROC: 0.8374962155616106
[[1177  291]
 [  98  253]]


F1 снизилась с 0,59 до 0,56.

### Модель случайного леса

In [27]:
predicted_test = best_model_RandomForestClassifier.predict(features_test)
probabilities_test = best_model_RandomForestClassifier.predict_proba(features_test)
probabilities_one_test = probabilities_test[:, 1]
print("F1:", f1_score(target_test, predicted_test), "AUC_ROC:", roc_auc_score(target_test, probabilities_one_test))
print(confusion_matrix(target_test,predicted_test))

F1: 0.6016042780748663 AUC_ROC: 0.8546698028986858
[[1296  172]
 [ 126  225]]


Только модель случайного леса достигла целевого значения F1 - 0,59. AUC_ROC составил 0,85

### Проверим модели на адекватность

In [29]:
dummy_results = []
dc = DummyClassifier(strategy = 'stratified', random_state = 12345)
dc.fit(features_train, target_train)
predicted_valid = dc.predict(features_valid)
probabilities_valid = dc.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
print("F1:", f1_score(target_valid, predicted_valid), "AUC_ROC:", roc_auc_score(target_valid, probabilities_one_valid))

F1: 0.20000000000000004 AUC_ROC: 0.494272508223548


"Простая модель" показала значительно меньшее значение F1, чем обученные модели. Поэтому ценность полученных моделей есть.

# ВЫВОД

Наилучшая модель - случайный лес - достигла порога целевой метрики F1 - 0.59 c результатом 0,6.